tagsets/orsee_mysql.php

```diff
@@ -247,5 +247,121 @@ function dump_array($array,$title="",$dolang=true) {
     echo '</TABLE>';
 }
 
+function check_database_upgrade() {
+    global $settings, $system__database_version;
+
+    if (!isset($settings['database_version'])) {
+        $settings['database_version']='2017112700';
+        $done=or_query("INSERT INTO ".table('options')." VALUES (1,'general',NULL,'database_version','".$settings['database_version']."')");
+    }
+    if ($settings['database_version']<$system__database_version) {
+        $done=or_upgrade_database();
+        return $done;
+    } else {
+        return false;
+    }
+}
+
+function upgrade_database_version($new_version) {
+    global $settings;
+    $settings['database_version']=$new_version;
+    $done=orsee_db_save_array(array('option_value'=>$new_version),'options',1,'option_id');
+    if(!$done) log__admin("Database upgrade error. Could not set database version to new version number ".$new_version."!");
+    return $done;
+}
+
+
+function or_upgrade_database() {
+    global $settings, $system__database_version, $system__database_upgrades;
+    
+    if (!isset($system__database_upgrades)) {
+        $system__database_upgrades=array();
+    }
+
+    // check $system__database_upgrades for syntax errors, resort database upgrades by version number
+    $database_upgrades=array(); $i=0; $continue=true;
+    foreach ($system__database_upgrades as $upgr) {
+        $i++;
+        if ($continue) {
+            $error=or_upgrade_check_syntax($upgr);
+            if ($error) {
+                $continue=false;
+                log__admin("Error in upgrade syntax for $system__database_upgrades item nb ".$i.": ".$error." Upgrade aborted.");
+            } else {
+                $upgr['item_nb']=$i;
+                $database_upgrades[$upgr['version']][]=$upgr;
+            }
+        }
+    }
+    if ($continue) {
+        $done=ksort($database_upgrades,SORT_NUMERIC);
+    }
+    
+    // run the updates, stop if there is an error
+    $continue=true;
+    foreach ($database_upgrades as $this_version=>$vupgrades) {
+        if ($this_version>$settings['database_version']) {
+            foreach ($vupgrades as $upgr) {
+                if ($continue) {
+                   if ($upgr['type']=='new_lang_item') {
+                        $done=lang__upgrade_symbol_if_not_exists($upgr['specs']);
```

Comment:
Shall we may be stop (`$continue=false;`) if any of those `$done` is false (should not happen, but just in case of some bug)?

Reply:
These upgrades may return false if the symbol already exists, but that would be no reason to stop. Similar for new rights.

Reply:
I see, then `$done` use in confusing, if it does not matter what return is, no variable is needed.

Reply:
I may be a little paranoid here, but I have the feeling that when I do not assign the result of a function to a variable, then that function may be executed in parallel (such that results of that function may not be relied upon). If I assign the result of a function to a variable, the execution has to wait until the function has been completed. But as said, that may just be superstition. ;-)